Cabal/src/Distribution/PackageDescription/Check.hs

```diff
@@ -91,33 +776,33 @@ data PackageCheck =
 
        -- | This package description is no good. There's no way it's going to
        -- build sensibly. This should give an error at configure time.
-       PackageBuildImpossible { explanation :: String }
+       PackageBuildImpossible { explanation :: CheckExplanation }
 
        -- | A problem that is likely to affect building the package, or an
        -- issue that we'd like every package author to be aware of, even if
        -- the package is never distributed.
-     | PackageBuildWarning { explanation :: String }
+     | PackageBuildWarning { explanation :: CheckExplanation }
 
        -- | An issue that might not be a problem for the package author but
        -- might be annoying or detrimental when the package is distributed to
        -- users. We should encourage distributed packages to be free from these
        -- issues, but occasionally there are justifiable reasons so we cannot
        -- ban them entirely.
-     | PackageDistSuspicious { explanation :: String }
+     | PackageDistSuspicious { explanation :: CheckExplanation }
 
        -- | Like PackageDistSuspicious but will only display warnings
        -- rather than causing abnormal exit when you run 'cabal check'.
-     | PackageDistSuspiciousWarn { explanation :: String }
+     | PackageDistSuspiciousWarn { explanation :: CheckExplanation }
 
        -- | An issue that is OK in the author's environment but is almost
        -- certain to be a portability problem for other environments. We can
        -- quite legitimately refuse to publicly distribute packages with these
        -- problems.
-     | PackageDistInexcusable { explanation :: String }
+     | PackageDistInexcusable { explanation :: CheckExplanation }
   deriving (Eq, Ord)
 
 instance Show PackageCheck where
-    show notice = explanation notice
+    show notice = ppExplanation (explanation notice)
```

Comment:
I don’t like `Show` instances that cannot be `Read` back, but since this is a public interface (used by `hackage-server`, `stack` etc.) I did not modify this behaviour.